provider/datagen/src/transform/cldr/datetime/neo.rs

```diff
@@ -513,6 +512,125 @@ fn datetimepattern_convert(
+    let pattern_long = pattern_str_long
+        .parse()
+        .expect("Failed to create a long Pattern from bytes.");
+    let pattern_medium = pattern_str_medium
+        .parse()
+        .expect("Failed to create a medium Pattern from bytes.");
+    let pattern_short = pattern_str_short
+        .parse()
+        .expect("Failed to create a short Pattern from bytes.");
+
+    let mut preferred_hour_cycle: Option<CoarseHourCycle> = None;
+    let arr = [
+        CoarseHourCycle::determine(&pattern_full),
+        CoarseHourCycle::determine(&pattern_long),
+        CoarseHourCycle::determine(&pattern_medium),
+        CoarseHourCycle::determine(&pattern_short),
+    ];
+    let iter = arr.iter().flatten();
+
+    if *langid == langid!("byn") || *langid == langid!("ssy") {
+        // byn uses H23H24 for full and H11H12 for the others
+        return CoarseHourCycle::H11H12;
+    }
+
+    for hour_cycle in iter {
+        if let Some(preferred_hour_cycle) = preferred_hour_cycle {
+            assert_eq!(
+                *hour_cycle, preferred_hour_cycle,
+                "Locale {langid} contained a mix of coarse hour cycle types"
+            );
+        } else {
+            preferred_hour_cycle = Some(*hour_cycle);
+        }
+    }
+
+    preferred_hour_cycle.expect("Could not find preferred hour cycle in locale {langid}")
+}
+
+// Time patterns have a manual implementation since they have custom supported_locales logic below
+impl DataProvider<TimePatternV1Marker> for DatagenProvider {
+    fn load(&self, req: DataRequest) -> Result<DataResponse<TimePatternV1Marker>, DataError> {
+        self.load_neo_patterns_key::<TimePatternV1Marker>(
+            req,
+            value!("gregory"),
+            timepattern_convert,
+        )
+    }
+}
+
+// Potential future optimization: if we split out aux subtags from supported_locales into a separate
+// method that eagerly generates all aux subtags (even if unused) and expects load() to figure out if the aux
+// subtag actually should be produced (by returning a special error), then this code is no longer necessary
+// and we can use a union of the H12/H24 key lengths arrays, instead checking for preferred hc
+// in timepattern_convert
+impl IterableDataProvider<TimePatternV1Marker> for DatagenProvider {
+    fn supported_locales(&self) -> Result<Vec<DataLocale>, DataError> {
+        let calendar = value!("gregory");
+        let mut r = Vec::new();
+
+        let cldr_cal = supported_cals()
+            .get(&calendar)
+            .ok_or_else(|| DataErrorKind::MissingLocale.into_error())?;
+        r.extend(self.cldr()?.dates(cldr_cal).list_langs()?.flat_map(|lid| {
+            let data = self
+                .load_calendar_dates(&lid, &calendar)
+                .expect("list_langs returned a language that couldn't be loaded");
+            let hc = hc_for(&data.time_formats, &lid);
+            let keylengths = match hc {
```

Comment:
small bug in this code: this should be matching on `hc.invert()` (or inverting the match, or inverting the constants)

The next commit obsoletes this, but noting this here in case we need to go back to this.